Support for Slack Interactive Messages

hatmarch opened the pull request:
This checkin adds the ability for the interceptor to deal with incoming messages from interactive slack messages (as if a tekton pipeline sends a message to slack).  See here for explanation of the case: https://api.slack.com/messaging/interactivity#components

See here for an example of how the interceptor is being used: https://github.com/hatmarch/tekton-comparison-demo/blob/k8-meetup-jul7/kube/tekton/triggers/slack/eventlistener-slack.yaml

This pull request also adds unobtrusive support for vscode remote (via .devcontainer)

See comments in each checkin message for more info

hatmarch committed:
Add VSCode Remote support

hatmarch committed:
Add support for "payload" parsing

Adds ability to support payloads that originate from Slack "Interactive Messages" (see here: https://api.slack.com/messaging/interactivity#components)

When Slack-Payload header is added, interceptor assumes that a "payload" element of the form points to pre-rendered JSON (that doesn't need to be remarshalled)

Adds a simple test for slack payload

Interceptor writes what it's about to return to standard out (to make debugging triggertemplates and bindings a bit easier

bigkevmcd commented on pkg/interception/handler.go:
```diff
@@ -59,6 +76,12 @@ func noFlatten(r *http.Request) bool {
+func payloadExtract(r *http.Request) bool {
+
+	return strings.ToLower(r.Header.Get(extractPayloadHeader)) == "true"
+	//return true
```
I think this can be tidied up.

hatmarch replied:
whoops.  Fixed

bigkevmcd commented on pkg/interception/handler.go:
```diff
@@ -33,6 +34,18 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", mimeJSON)
+
+	if payloadExtract(r) {
+		myData := r.PostForm.Get("payload")
```
What do you think about naming this `payload` instead of `myData` ?

Also, I think this should really be parsing this as JSON (`map[string]interface{}`) and returning it inside a map with the `prefixFromRequest` rather than generating a string here?

hatmarch replied:
I've updated the whole flow, see changes (and below)

bigkevmcd reviewed (changes requested):
Thank you very much for the PR, looks mostly good to me, and I'd love to see the outcome of the comparison.

I've got some ideas for some bits that might make it a bit more robust.

If there's anything else I can do to help with your talk, I'd be happy to :-)

hatmarch committed:
Reuse existing marshaling path

Update the handler flow to mark the incoming slack interaction payload as JSON data already to keep it from being re-marshalled in the call to json.Marshal.  This allows greater reuse work common to this and existing flows (e.g. adding the prefix or printing the message to stdout before returning)

hatmarch committed:
Remove dead code comment

hatmarch commented:
Hi Kevin,
Totally agree with your comments.  This is my first exposure to golang so I was just trying to get something working quickly and to raise the functionality to your attention.

Regardless, I've just discovered `json.RawMessage` which has allowed me to follow the existing flow for better readability and less repetition of code.  I think I've resolved your comments now, let me know if you agree.

Re: talk> nothing right now thanks, but you might want to take a look at the Tekton aspects of the repo I referenced [here](https://github.com/hatmarch/tekton-comparison-demo/tree/k8-meetup-jul7) and let me know if I'm missing anything.  One thing that I find a little annoying in the current version of Tekton referenced by OpenShift Pipelines is that parallel tasks that share a workspace are not assigned to the same node which causes intermittent failures when the workspace can only be set as `ReadWriteOnce`

Cheers,
Marc

bigkevmcd reviewed (approved):
Great, thanks for this.
